astropy/time/core.py

```diff
@@ -1236,12 +1234,15 @@ def argsort(self, axis=-1):
         is used, and that corresponding attributes are copied.  Internally,
         it uses :func:`~numpy.lexsort`, and hence no sort method can be chosen.
         """
-        jd_approx = self.jd
-        jd_remainder = (self - self.__class__(jd_approx, format='jd', scale=self.scale)).jd
```

Comment:
Do you recall why it was written this way? That is likely the performance bottleneck here. I can't think of why it should be done this way, but sometimes the original coder (even when it was yourself) understood something that has since been forgotten. :smile:

Reply:
Indeed! This line has been here from the start (#3681), and `argmin` and `argmax` were treated differently already too. I have no recollection why at the time I did not think the same logic that is written down for `argmin` might not hold for `argsort`. Perhaps it is that at the time we did not as strictly guarantee that jd1 would be exact (being integer + 0.5)? In any case, because jd1 is exactly representable, this means there is no point going through initializing ourselves and doing the subtraction that way; one can just write it out like for `argmin` and realize that `jd1-approx` cannot loose bits, and the subsequent `+jd2` cannot loose bits either. Indeed, if we completely trusted that `jd1, jd2` had absolutely always gone through `day_frac`, we should be able to sort on `jd1, jd2`. But I'm just a bit wary that rounding on +/-0.5 can sometimes throw a spanner in the works. 

Anyway, it is right to be worried, so I wrote an extra test for `test_precision` using hypothesis, which will warn us if things fail (the existing tests are actually quite good, I think, but fine to be safe here!)